packages/cloud_functions/cloud_functions_platform_interface/lib/cloud_functions_platform_interface.dart

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+library cloud_functions_platform_interface;
+
+import 'dart:async';
+
+import 'package:firebase_core/firebase_core.dart';
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart' show required, visibleForTesting;
+
+part 'src/https_callable.dart';
+part 'src/https_callable_result.dart';
+part 'src/method_channel_cloud_functions.dart';
+
+/// The interface that implementations of `cloud_functions` must extend.
+///
+/// Platform implementations should extend this class rather than implement it
+/// as `cloud_functions` does not consider newly added methods to be breaking
+/// changes. Extending this class (using `extends`) ensures that the subclass
+/// will get the default implementation, while platform implementations that
+/// `implements` this interface will be broken by newly added
+/// [CloudFunctionsPlatform] methods.
+abstract class CloudFunctionsPlatform {
+  /// Only mock implementations should set this to `true`.
+  ///
+  /// Mockito mocks implement this class with `implements` which is forbidden
+  /// (see class docs). This property provides a backdoor for mocks to skip the
+  /// verification that the class isn't implemented with `implements`.
+  @visibleForTesting
+  bool get isMock => false;
+
+  /// The default instance of [CloudFunctionsPlatform] to use.
+  ///
+  /// Platform-specific plugins should override this with their own class
+  /// that extends [CloudFunctionsPlatform] when they register themselves.
+  ///
+  /// Defaults to [MethodChannelCloudFunctions].
+  static CloudFunctionsPlatform get instance => _instance;
+
+  static CloudFunctionsPlatform _instance = MethodChannelCloudFunctions();
+
+  // TODO(amirh): Extract common platform interface logic.
+  // https://github.com/flutter/flutter/issues/43368
+  static set instance(CloudFunctionsPlatform instance) {
+    if (!instance.isMock) {
+      try {
+        instance._verifyProvidesDefaultImplementations();
+      } on NoSuchMethodError catch (_) {
+        throw AssertionError(
+            'Platform interfaces must not be implemented with `implements`');
+      }
+    }
+    _instance = instance;
+  }
+
+  /// This method ensures that [CloudFunctionsPlatform] isn't implemented with `implements`.
+  ///
+  /// See class docs for more details on why using `implements` to implement
+  /// [CloudFunctionsPlatform] is forbidden.
+  ///
+  /// This private method is called by the [instance] setter, which should fail
+  /// if the provided instance is a class implemented with `implements`.
+  void _verifyProvidesDefaultImplementations() {}
+
+  /// Gets an instance of a Callable HTTPS trigger in Cloud Functions.
```

Comment:
From looking at the `cloud_functions` package, I think the correct platform interface is actually:

```dart
dynamic callCloudFunction({
  @required String appName,
  @required String functionName, 
  String region,
  String origin,
  Duration timeout,
  dynamic parameters,
}) { ... }
```

Platform interfaces should strive to be very simple abstraction of the `MethodChannel` calls that already exist in the current plugin unless the plugin is very complex. The idea is that this is the minimal interface that a platform would need to implement to support this plugin. By pulling in the abstractions from the existing plugin (`HttpsCallable`, `HttpsCallableResult`) you are making `package:cloud_functions` less flexible.

If you just have `dynamic callCloudFunction(...)` then in `package:cloud_functions` you just replace the method call with `CloudFunctionPlatform.instance.callCloudFunction(...)` and the rest of the plugin stays the same. The plugin is also free to change its `HttpsCallable` abstraction however it wants, since the platform interface offers the basic functionality.

Reply:
Okay, that makes sense. However, wouldn't it also be good to include the `useFunctionsEmulator` call? After all, the Firebase CLI includes a local emulator for functions specifically for development; wouldn't it make sense for the platform interface to expose a control to allow code to specify whether to use the emulator or the real functions, no matter what platform implementation is being used? Or does this somehow belong on the `cloud_functions` plugin?

Reply:
I guess my concern is that this would be a breaking change. Maybe that's okay, given that the point here is to prepare for flutter_web. Do you reckon that's acceptable?

Reply:
It wouldn't be a breaking change if you leave `HttpsCallable`, `HttpsCallableResult`, etc in `package:cloud_functions`. As for `useFunctionsEmulator`, you can document that setting an `origin` will change it to use a functions emulator. See the Android code here: https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_functions/android/src/main/java/io/flutter/plugins/firebase/cloudfunctions/CloudFunctionsPlugin.java#L49